src/java.base/share/classes/sun/security/pkcs12/PKCS12KeyStore.java

```diff
@@ -1510,8 +1513,9 @@ private byte[] calculateMac(char[] passwd, byte[] data)
             byte[] macResult = m.doFinal();
 
             // encode as MacData
-            MacData macData = new MacData(algName, macResult, salt,
-                    macIterationCount);
+            macData = new MacData(algName, macResult, params,
+                    defaultMacAlgorithm().replace("PBE", ""),
+                    extraSalt, extraIterationCount);
```

Comment:
I'm not sure why you need the additional parameters (`params`, `extraSalt`, `extraIterationCount`). I think (and I might be wrong), you can use the existing constructor and inside MacData, detect if PBMAC1 is being used by checking the algorithm name, and then encode the salt and iteration count in the PBMAC1 algorithm parameters as per RFC 9579.

Reply:
I took your suggestion about detecting PBMAC1 via the algorithm name. Thanks. It works.
The `extra*` parameters are necessary because RFC 9579 says they have to be present.

   |  c.  If the PBMAC1 algorithm is used, the macSalt value MUST be
   |      ignored.  For backwards compatibility, it SHOULD NOT be empty.
   |  
   |  d.  If the PBMAC1 algorithm is used, the iterations value MUST be
   |      ignored.  For backwards compatibility, it SHOULD have a non-
   |      zero positive value.

Basically a PBMAC1 protected keystore contains an unused old style MAC at the end. If upgrading an old keystore to a PBMAC1 protected keystore, then it seems better to use values from the old MAC rather than make them up.